node/network/collator-protocol/src/collator_side/validators_buffer.rs

```diff
@@ -0,0 +1,287 @@
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Polkadot.  If not, see <http://www.gnu.org/licenses/>.
+
+//! Validator groups buffer for connection managements.
+//!
+//! Solves 2 problems:
+//! 	1. A collator may want to stay connected to multiple groups on rotation boundaries.
+//! 	2. It's important to disconnect from validator when there're no collations to be fetched.
+//!
+//! We keep a simple FIFO buffer of N validator groups and a bitvec for each advertisement,
+//! 1 indicating we want to be connected to i-th validator in a buffer, 0 otherwise.
+//!
+//! The bit is set to 1 on new advertisements, and back to 0 when a collation is fetched
+//! by a validator.
+//!
+//! The bitwise OR over known advertisements gives us validators indices for connection request.
+
+use std::{
+	collections::{HashMap, VecDeque},
+	num::NonZeroUsize,
+	ops::Range,
+};
+
+use bitvec::{bitvec, vec::BitVec};
+
+use polkadot_primitives::v2::{AuthorityDiscoveryId, GroupIndex, Hash, SessionIndex};
+
+/// The ring buffer stores at most this many unique validator groups.
+///
+/// This value should be chosen in way that all groups assigned to our para
+/// in the view can fit into the buffer.
+pub const VALIDATORS_BUFFER_CAPACITY: NonZeroUsize = match NonZeroUsize::new(3) {
+	Some(cap) => cap,
+	None => panic!("buffer capacity must be non-zero"),
+};
+
+/// Unique identifier of a validators group.
+#[derive(Debug)]
+struct ValidatorsGroupInfo {
+	len: usize,
+	session_index: SessionIndex,
+	group_index: GroupIndex,
+}
+
+/// Ring buffer of validator groups.
+///
+/// Tracks which peers we want to be connected to with respect to advertised collations.
+#[derive(Debug)]
+pub struct ValidatorGroupsBuffer {
+	buf: VecDeque<ValidatorsGroupInfo>,
+	validators: VecDeque<AuthorityDiscoveryId>,
+	should_be_connected: HashMap<Hash, BitVec>,
+
+	cap: NonZeroUsize,
+}
+
+impl ValidatorGroupsBuffer {
+	/// Creates a new buffer with a non-zero capacity.
+	pub fn with_capacity(cap: NonZeroUsize) -> Self {
+		Self {
+			buf: VecDeque::new(),
+			validators: VecDeque::new(),
+			should_be_connected: HashMap::new(),
+			cap,
+		}
+	}
+
+	/// Returns discovery ids of validators we have at least one advertised-but-not-fetched
+	/// collation for.
+	pub fn validators_to_connect(&self) -> Vec<AuthorityDiscoveryId> {
+		let validators_num = self.validators.len();
+		let bits = self
+			.should_be_connected
+			.values()
+			.fold(bitvec![0; validators_num], |acc, next| acc | next);
+
+		self.validators
+			.iter()
+			.enumerate()
+			.filter_map(|(idx, authority_id)| bits[idx].then_some(authority_id.clone()))
+			.collect()
+	}
+
+	/// Note a new collation distributed, setting this group validators bits to 1.
+	///
+	/// If max capacity is reached and the group is new, drops validators from the back
+	/// of the buffer.
+	pub fn note_collation_distributed(
```

Comment:
We call it when distribute a collation to group, not when advertising to peer.

Reply:
But the calling function is only called `distribute`, because it initiates distribution - by advertising. By the time we call `note_collation_distributed` - we only initiated advertising. I find this especially confusing, because once `distributed` ... which implies validators have fetched it, I would expect bits to get cleared not set.

Reply:
agreed